Sources/Datadog/RUM/RUMVitals/VitalCPUReader.swift

```diff
@@ -0,0 +1,88 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+import UIKit.UIApplication
+
+/// A class reading the CPU ticks of the processor.
+internal class VitalCPUReader: VitalReader {
+    /// host_cpu_load_info_count is 4 (tested in iOS 14.4)
+    private static let host_cpu_load_info_count = MemoryLayout<host_cpu_load_info>.stride / MemoryLayout<integer_t>.stride
+    private var totalInactiveTicks: natural_t = 0
+    private var utilizedTicksWhenResigningActive: natural_t? = nil
+
+    init(notificationCenter: NotificationCenter = .default) {
+        notificationCenter.addObserver(self, selector: #selector(appWillResignActive), name: UIApplication.willResignActiveNotification, object: nil)
+        notificationCenter.addObserver(self, selector: #selector(appDidBecomeActive), name: UIApplication.didBecomeActiveNotification, object: nil)
+    }
+
+    func readVitalData() -> Double? {
+        if let ticks = readUtilizedTicks() {
+            let ongoingInactiveTicks = ticks - (utilizedTicksWhenResigningActive ?? ticks)
+            let inactiveTicks = totalInactiveTicks + ongoingInactiveTicks
+            return Double(ticks - inactiveTicks)
+        }
+        return nil
+    }
+
+    @objc
+    private func appWillResignActive() {
+        utilizedTicksWhenResigningActive = readUtilizedTicks()
```

Comment:
When I looked further, I can see that this will lead to threading issue. The `readVitalData()` will be accessed from background thread (by `RUMScope`), but the `appWillResignActive()` will be called on the main thread. Both access the `utilizedTicksWhenResigningActive` state by either reading or writing.

We can solve it later, when actually plugging `VitalReaders` to RUM (this will pop up in threading tests), but it's worth leaving this note.

Reply:
yes, that's a good point and i'm still not sure where the `VitalReader`s will live.
if they live in `RUMMonitor` we can read data in main thread within `rum.start/stopView(...)` methods (even before `process(command:)`) and that probably gives the most accurate measurement.
if they live in the scopes then we will need to add extra code as you said.

i'm leaving a comment 👍